### ISSUE TITLE
Graceful degradation of error logging

### ISSUE THREAD

**Issue by alexlatchford** [MindscapeHQ/raygun4py#49]
Hi all,

So I was just tailing our production logs and found a few of these errors:

```
172.17.42.1 - - [06/Oct/2016:12:29:45] "" 200 - "" ""
Exception in thread Thread-8:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/threading.py", line 811, in __bootstrap_inner
    self.run()
  File "/usr/lib64/python2.7/threading.py", line 764, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/var/www/a8_product/database/__init__.py", line 83, in _connection
    return f(*args, **kwargs)
  File "/var/www/a8_product/controllers/briefs/refresh.py", line 121, in _refresh_brief_articles
    populate_brief(db_brief, event=event)
  File "/var/www/a8_product/controllers/briefs/refresh.py", line 29, in populate_brief
    audit_brief(db_brief, event)
  File "/var/www/a8_product/controllers/briefs/audit.py", line 15, in audit_brief
    raygun_log(e.message, custom_data={'brief_uuid': str(db_brief.uuid)})
  File "/var/www/a8_product/lib/raygun_logging.py", line 10, in raygun_log
    return _raygun_log(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/hg_logging/raygun/logger.py", line 34, in raygun_log
    provider.sender.send_exception(**kwargs)
  File "/usr/lib/python2.7/site-packages/raygun4py/raygunprovider.py", line 106, in send_exception
    message = self._create_message(errorMessage, tags, custom_data, http_request, extra_environment_data)
  File "/usr/lib/python2.7/site-packages/raygun4py/raygunprovider.py", line 131, in _create_message
    .set_environment_details(extra_environment_data) \
  File "/usr/lib/python2.7/site-packages/raygun4py/raygunmsgs.py", line 35, in set_environment_details
    "architecture": platform.architecture()[0],
  File "/usr/lib64/python2.7/platform.py", line 1111, in architecture
    output = _syscmd_file(executable, '')
  File "/usr/lib64/python2.7/platform.py", line 1054, in _syscmd_file
    stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
  File "/usr/lib64/python2.7/site-packages/gevent/subprocess.py", line 534, in __init__
    reraise(*exc_info)
  File "/usr/lib64/python2.7/site-packages/gevent/subprocess.py", line 503, in __init__
    restore_signals, start_new_session)
  File "/usr/lib64/python2.7/site-packages/gevent/subprocess.py", line 1112, in _execute_child
    self.pid = fork_and_watch(self._on_child, self._loop, True, fork)
  File "/usr/lib64/python2.7/site-packages/gevent/os.py", line 341, in fork_and_watch
    watcher = loop.child(pid, ref=ref)
  File "gevent/corecext.pyx", line 511, in gevent.corecext.loop.child (gevent/gevent.corecext.c:9881)
  File "gevent/corecext.pyx", line 1876, in gevent.corecext.child.__init__ (gevent/gevent.corecext.c:32615)
TypeError: child watchers are only available on the default loop
```

This particular trace is from a background thread, launched as a part of a web request which altered an object which we audit out S3. Obviously have a bug somewhere that is causing that process to fail, now I see this error is stopping the original bug being recorded in raygun 👎 

I'm in a bit of bind here, looks like the gevent bug (see warning at the top: http://www.gevent.org/gevent.subprocess.html) is a limitation that won't be fixed. I'm trying porting our background threads over to gevent but I figured I'd suggest a new feature too.

So we never really use the architecture stuff when debugging and I'd much rather have the error with less detail than no error whatsoever. So can I ask for either the ability to turn off/on feature enrichment manually and/or to have it gracefully fallback so it just get's the minimum it needs but can still log the error.

Cheers,
Alex

**Comment by fundead** [comment]
Thanks for reporting this one, if possible I'd like to get the provider tightened up for your use case here. If I've read the stack and scenario correctly the raygunmsgs.py platform.architecture()[0] call is failing, and you'd like a best-effort approach for getting the environment data so that it doesn't result in unhandled exceptions - is the right?

If that's the case one approach would be to guard `set_environment_details` with an except, this approach is used in several of our other providers as access to environment data varies wildly between sub-frameworks. This would surface the safe properties first so it should degrade nicely and send the available env data without spamming your logs. Does this sound like what you're looking for?

**Comment by alexlatchford** [comment]
Yes I think that's the case. I'm advocating for always raising an error to raygun, even if it's skinnier than normal rather than dropping the error to the log file. 

Thanks, Alex

**Comment by fundead** [comment]
Great, yeah that sounds like the right sort of behavior we should have. I'll ensure that a work item for this is put on the roadmap and fixed in the near term for you.

**Comment by alexlatchford** [comment]
Thanks, what kind of an ETA are we looking at for this? Days/Weeks/Months? 

Got a couple of pretty painful errors I need to work around if this is going to be a while.

**Comment by fundead** [comment]
I'll get a release out for you today containing a fix for this (next few hours).

**Comment by alexlatchford** [comment]
Great thanks Callum

**Comment by alexlatchford** [comment]
Hey Callum, any news on this one?

**Comment by fundead** [comment]
Hi Alex,

Unfortunately yesterday was a public holiday here so I've only just been able to get back into the office and finalize the PR. This is now complete and has been released as v3.1.3 on PyPI. Hope this fixes the noise in your logs and you continue to find Raygun useful.

Regards,

Callum Gavin
Raygun Limited

**Comment by alexlatchford** [comment]
No problems Callum, thanks for the quick turn around. Really appreciate the effort! 😄
